module/zfs/vdev_raidz.c

```diff
@@ -2281,6 +2288,123 @@ vdev_raidz_min_asize(vdev_t *vd)
 	    vd->vdev_children);
 }
 
+/*
+ * return B_TRUE if a read should be skipped due to being too slow.
+ *
+ * In vdev_child_slow_outlier() it looks for outliers based on disk
+ * latency from the most recent child reads.  Here we're checking if,
+ * over time, a disk has has been an outlier too many times.
+ */
+boolean_t
+vdev_skip_latency_outlier(vdev_t *vd, zio_flag_t io_flags)
+{
+	if (raid_read_sit_out_secs == 0)
+		return (B_FALSE);
+
+	/* Avoid skipping a data column read when resilvering */
+	if (io_flags & (ZIO_FLAG_SCRUB | ZIO_FLAG_RESILVER))
+		return (B_FALSE);
+
+	/*
+	 * Ignore until we have a reasonable number of outlier events.
+	 * This is the expected exit in most cases.
+	 */
+	if (atomic_load_64(&vd->vdev_outlier_count) < 50)
+		return (B_FALSE);
+
+	vdev_t *raidvd = vd->vdev_parent;
+
+	/*
+	 * We're using the stat lock to also synchronize access to the
+	 * vdev_read_sit_out_expire states. This code path is limited
+	 * to an identified outlier vdev.
+	 */
+	mutex_enter(&vd->vdev_stat_lock);
```

Comment:
A separate RW lock would be better for this.  ZFS has a checkered past with overloading existing locks ... https://github.com/openzfs/zfs/pull/16507